internal/db/merge.go

```diff
@@ -0,0 +1,364 @@
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+
+	"github.com/sourcenetwork/corelog"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/events"
+	"github.com/sourcenetwork/defradb/internal/core"
+	coreblock "github.com/sourcenetwork/defradb/internal/core/block"
+	"github.com/sourcenetwork/defradb/internal/db/base"
+	"github.com/sourcenetwork/defradb/internal/merkle/clock"
+	merklecrdt "github.com/sourcenetwork/defradb/internal/merkle/crdt"
+)
+
+func (db *db) handleMerges(ctx context.Context, merges events.Subscription[events.DAGMerge]) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case merge, ok := <-merges:
+			if !ok {
+				return
+			}
+			go func() {
+				err := db.executeMerge(ctx, merge)
+				if err != nil {
+					fmt.Println(err)
+					log.ErrorContextE(
+						ctx,
+						"Failed to execute merge",
+						err,
+						corelog.String("cid", merge.Cid.String()),
+					)
+				}
+			}()
+		}
+	}
+}
+
+func (db *db) executeMerge(ctx context.Context, dagMerge events.DAGMerge) error {
+	defer func() {
+		// Notify the caller that the merge is complete.
+		if dagMerge.MergeCompleteChan != nil {
+			close(dagMerge.MergeCompleteChan)
+		}
+	}()
+	ctx, txn, err := ensureContextTxn(ctx, db, false)
+	if err != nil {
+		return err
+	}
+	defer txn.Discard(ctx)
+	mp, err := db.newMergeProcessor(ctx, dagMerge.Cid, dagMerge.SchemaRoot)
+	if err != nil {
+		return err
+	}
+	mt, err := mp.getHeads(ctx)
+	if err != nil {
+		return err
+	}
+	err = mp.getComposites(ctx, dagMerge.Cid, mt)
+	if err != nil {
+		return err
+	}
+	err = mp.merge(ctx)
+	if err != nil {
+		return err
+	}
+	err = mp.syncIndexedDocs(ctx)
+	if err != nil {
+		return err
+	}
+	return txn.Commit(ctx)
+}
+
+type mergeProcessor struct {
+	ctx              context.Context
+	txn              datastore.Txn
+	ls               linking.LinkSystem
+	docID            client.DocID
+	mCRDTs           map[uint32]merklecrdt.MerkleCRDT
+	col              *collection
+	schemaVersionKey core.CollectionSchemaVersionKey
+	dsKey            core.DataStoreKey
+	composites       *list.List
+}
+
+func (db *db) newMergeProcessor(ctx context.Context, cid cid.Cid, rootSchema string) (*mergeProcessor, error) {
+	txn, ok := TryGetContextTxn(ctx)
+	if !ok {
+		return nil, ErrNoTransactionInContext
+	}
+
+	ls := cidlink.DefaultLinkSystem()
+	ls.SetReadStorage(txn.DAGstore().AsIPLDStorage())
+	nd, err := ls.Load(linking.LinkContext{Ctx: ctx}, cidlink.Link{Cid: cid}, coreblock.SchemaPrototype)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := coreblock.GetFromNode(nd)
+	if err != nil {
+		return nil, err
+	}
+
+	cols, err := db.getCollections(
+		ctx,
+		client.CollectionFetchOptions{
+			SchemaRoot: immutable.Some(rootSchema),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	col := cols[0].(*collection)
```

Comment:
todo: Please document this `[0]`: why it is okay to do so now, why it is safe (and we are sure that there will be at least one), and why it might need to change in the future.

Reply:
I added a comment. Let me know what you think.

Reply:
LGTM, thanks :)